### DISCUSSION TITLE
Add a RequestPath() function

### DISCUSSION THREAD

**Discussion by viccon** [earthboundkid/requests#106] [Ideas] [answered]
Hi,

I've recently been working with [toxiproxy](https://github.com/Shopify/toxiproxy) to simulate failures and high latency for upstream systems. In order to make that work, I had to change the `Host` field of the request:

```go
var response map[string]Episode
requestBuilder := requests.URL(c.baseURL).
	Path("/episodes").
	Param("ids", strings.Join(ids, ",")).
	Client(c.httpClient).
	Headers(c.headers).
	ToJSON(&response)

request, err := requestBuilder.Request(timeoutCtx)
if err != nil {
	return response, err
}

request.Host = c.host
err = requestBuilder.Do(request)
```

What do you think about adding a `RequestHost` function to the `RequestBuilder`? 

E.g:

```go
var response map[string]Episode
err := requests.URL(c.baseURL).
	Path("/episodes").
	Param("ids", strings.Join(ids, ",")).
	Headers(c.headers).
	ToJSON(&response).
	RequestHost(c.host).
	Fetch(ctx)
```

I tried using the current `Host` function, but it sets the host field on the requests url rather than the request itself. If you think having this functionality would be valuable I'd be happy to implement it.

Thank you for creating this library it's great!

**Comment by earthboundkid** [comment, marked as answer]
`.Host` sets the host on the URL, which in turns gets passed into the Request. Is there a reason you need the host to be only on the request and not the URL? I know there's an option for that in curl, but I've only ever needed it because I was testing whether a server would work even if it got a bad host in.

I think the workaround for now would be to make a custom transport like

```go
func WithHost(rt requests.Transport, s string) requests.Transport {
	return requests.RoundTripFunc(func(req *http.Request) (res *http.Response, err error) {
		rt = cmp.Or(rt, http.DefaultTransport)
		req2 := *req
		req2.Host = s
		return rt.RoundTrip(&req2)
	})
}
```

**Reply by earthboundkid** [reply]
Even simpler, I guess you could set the Host header and then maybe Go won't add it back in? Haven't tested that.

**Reply by earthboundkid** [reply]
In some quick testing, if you just set the Host header, but not the Request.Host, then Go ignores the Host header and uses the URL.Host. So I guess you have to use the WithHost transport to make it work. I'm not sure this is common enough to go in the main package. It's great for the Wiki though.

**Reply by viccon** [reply]
Yep, the host header alone doesn't work because Go ignores it, and if I use the `Host` function from the library I end up not going through the proxy because it changes the url.

I'm essentially trying to achieve the `--connect-to` option from curl.

Thank's a lot for your time and for that snippet. I'll give it a go tomorrow and let you know how it goes!

**Reply by viccon** [reply]
Sorry, I forgot to update this thread! Your suggestion did exactly what I needed. I never used a roundtripper before, but it's ability to modify the request before its sent, along with setting the `ServerName` on the tls config to the same value, made this work without breaking the chain. Thank you so much 🙏

I still wasn't able to understand why Go doesn't check if the host header is set, before it takes the value from the URL. It seem a bit strange to me that it would override your headers.
